### ISSUE TITLE
Virtual DOM API Proposal

### ISSUE THREAD

**Issue by the-dr-lazy** [the-dr-lazy/purescript-monarch#5]
**TL;DR:** The RFC has been accomplished. **Flatten Attributes and Outputs With Nested Optional Properties and Hooks** API has been accepted and I'm gonna to implement it. Special thanks to @jedimahdi.

## RFC: Virtual DOM API

Meanwhile I'm working on Monarch elements, I'm thinking to what is the best option for virtual DOM API?!
In the following I propose multiple ways to represent virtual DOM in PureScript for Monarch. Please let me know your opinions about each one and which one you mostly prefer.

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Separates type of the data (attributes, properties, etc...).
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hook: { pre: const Pre }
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't separate type of the data and it can lead to name clashes as there are lots of properties and attributes with similar names in the HTML specification (property, attribute, output (event), hook, etc...)
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hookPre: const Pre
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Separates type of the data (attributes, properties, etc...).
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Boilerplate
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { attrs:   mkAttrs  { href: "/" }
         , outputs: mkOutput { click: const Click
                             , blur: const Blur
                             }
         , hooks:   mkHooks  { pre: const Pre }
         }
         [ Sphoenoid.slider { attrs:   mkAttrs   { step: 1 }
                            , outputs: mkOutputs { next: const Next }
                            }
                            [ text "!!!" ]
         ]
```

#### Pros
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Separates type of the data (attributes, properties, etc...).
- Impossible to declare an attribute or a property multiple times.

#### Cons
- Hard to implement: It needs some type-level programming features in PureScript.
- Doesn't support multiple output handlers.

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur:  const Blur
         , hooks:   mkHooks  { pre: const Pre }
         }
         [ Sphoenoid.slider { step: 1 
                            , onNext: const Next
                            }
                            [ text "!!!" ]
         ]
```

#### Pros
- Easy to implement
- Extremely type-safe: it can specify each HTML element handles which attributes, properties, etc... .
- Already extensible: developers can define their own attribute, property, output, hook, etc... .
- Supports multiple output handlers.
- Separates type of the data (attributes, properties, etc...).

#### Cons
- Possible to declare an attribute or a property multiple times.

```purs
view :: VirtualNode Message
view = a [ Attr.href "/"
         , Output.click $ const Click
         , Output.blur  $ const Blur
         , Hook.pre     $ const Pre
         ]
         [ Sphenoid.slider [ Attr.step 1
                           , Output.next $ const Next
                           ]
                           [ text "!!!" ]
         ]
```

**Comment by jedimahdi** [comment]
I prefer "Flat Record" in a way that some of props (like hooks) can be nested for example


```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hook:  { pre: const Pre }
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```


"Array and Modules" seems pretty flexible but I think it can cause bugs and since all data types will be arrays, view function can be confusing.

**Comment by the-dr-lazy** [comment]
> I prefer "Flat Record" in a way that some of props (like hooks) can be nested for example

~~Actually it's not possible to have nested optional record in PureScript yet (based on my experience)~~. Do you have any idea to make it possible?

**Comment by jedimahdi** [comment]
It's possible to have nested optional record if we write previous example like this 
```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hook: makeHook { pre: const Pre }
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```
but `makeHook` seems annoying.

**Comment by the-dr-lazy** [comment]
> It's possible to have nested optional record

It's not nested optional record but it makes a way to represent something similar. BTW, very good suggestion. I will add it to the main comment of this issue as an option.

**Comment by jedimahdi** [comment]
I don't know it's possible or not but can we have array of hook modules in hook field ? something like this:

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hook: [ Hook.pre $ const Pre ]
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next3
                           }
                           [ text "!!!" ]
         ]
```

**Comment by the-dr-lazy** [comment]
> I don't know it's possible or not but can we have array of hook modules in hook field

It's possible! Also it's possible for outputs too.

**Comment by the-dr-lazy** [comment]
It's possible to make **Array and Modules** approach completely type-safe.

**Comment by the-dr-lazy** [comment]
> Actually it's not possible to have nested optional record in PureScript yet (based on my experience). Do you have any idea to make it possible?

Found a way to make nested optional record possible in PureScript! :D 
Which means we are able to make following virtual DOM API possible:

```purs
view :: VirtualNode Message
view = a { href: "/"
         , onClick: const Click
         , onBlur: const Blur
         , hook:  { pre: const Pre }
         }
         [ Sphenoid.slider { step: 1
                           , onNext: const Next
                           }
                           [ text "!!!" ]
         ]
```
